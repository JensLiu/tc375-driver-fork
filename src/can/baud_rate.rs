// Many integer and float conversions are done in this file, we want to get rid of them
#![deny(clippy::as_conversions, clippy::float_cmp)]

use crate::util::F32Abs;

pub(super) const NBTP_NBRP_MSK: usize = 0x1ff;
pub(super) const NBTP_NTSEG1_MSK: usize = 0xff;
pub(super) const NBTP_NTSEG2_MSK: usize = 0x7f;

pub(super) const DBTP_DBRP_MSK: usize = 0x1f;
pub(super) const DBTP_DTSEG1_MSK: usize = 0x1f;
pub(super) const DBTP_DTSEG2_MSK: usize = 0xf;

pub(super) struct BestBaudRate {
    pub(super) tbaud: i32,
    pub(super) brp: i32,
}

pub(super) fn get_best_baud_rate(
    brp_msk: usize,
    tseg1_msk: usize,
    tseg2_msk: usize,
    module_freq: f32,
    baudrate: u32,
) -> BestBaudRate {
    /* search for best baudrate */

    let max_brp = brp_msk as i32 + 1;
    let min_brp = 1;

    let max_tbaud = (tseg1_msk as i32 + 1) + (tseg2_msk as i32 + 1) + 1;
    let min_tbaud = 8;

    let mut best_error = baudrate as f32;
    let mut best_brp = 1;
    let mut best_tbaud = 8;
    let mut tmp_brp = 1;
    let mut tmp_tbaud: i32 = 0;

    while tmp_brp <= max_brp {
        let f_quanta = module_freq / tmp_brp as f32;
        tmp_tbaud = (f_quanta / baudrate as f32) as _;

        if tmp_tbaud == 0 {
            break; /* to avoid division by 0 */
        }

        let temp_baudrate = f_quanta / tmp_tbaud as f32;
        let error = (temp_baudrate - baudrate as f32).abs();

        if tmp_tbaud < min_tbaud {
            break; /* below the minimum allowed limits, break is required otherwise TSEG1 and TSEG2 may result in negitive values */
        }

        if (tmp_tbaud <= max_tbaud) && (best_error >= error) {
            best_brp = tmp_brp;
            best_tbaud = tmp_tbaud;
            best_error = error;

            if (tmp_tbaud <= 20) && (error < 0.1) {
                break; /* optimal condition */
            }
        }
        tmp_brp += 1;
    }

    if (best_brp == 0) && (tmp_brp == (max_brp + 1)) {
        // force max brp and tbaud
        best_brp = max_brp;
        best_tbaud = max_tbaud;
    }

    if (best_brp == 0) && (tmp_tbaud < min_tbaud) {
        // force min brp and tbaud
        best_brp = min_brp;
        best_tbaud = min_tbaud;
    }

    BestBaudRate {
        tbaud: best_tbaud,
        brp: best_brp,
    }
}

pub(super) fn get_best_sample_point(
    tseg1_msk: usize,
    tseg2_msk: usize,
    best_tbaud: i32,
    sample_point: u16,
) -> (i32, i32) {
    /* search for best sample point */

    let mut best_error = sample_point as f32 * 0.25; /* 25% tolerance in sample point as max error */
    let max_tseg1 = tseg1_msk as i32 + 1;
    let min_tseg1 = 3;
    let max_tseg2 = tseg2_msk as i32 + 1;
    let min_tseg2 = 2;

    let max_tseg1 = if best_tbaud < max_tseg1 {
        best_tbaud
    } else {
        max_tseg1
    };
    let mut best_tseg1 = max_tseg1;

    for temp_tseg1 in (min_tseg1..=max_tseg1).rev() {
        let temp_sample_point = ((temp_tseg1 + 1) * 10000) / best_tbaud;
        let error = temp_sample_point - sample_point as i32;
        let error = if error < 0 { -error } else { error };

        if best_error > error as _ {
            best_tseg1 = temp_tseg1;
            best_error = error as _;
        }

        if temp_sample_point < sample_point as _ {
            /*least possible error */
            break; /* least possible error has already occured */
        }
    }

    let best_tseg2 = if (best_tbaud - best_tseg1 - 1) > max_tseg2 {
        // force max tseg2
        max_tseg2
    } else {
        best_tbaud - best_tseg1 - 1
    };

    let best_tseg2 = if best_tseg2 < min_tseg2 {
        // force min tseg2
        min_tseg2
    } else {
        best_tseg2
    };

    (best_tseg1, best_tseg2)
}

pub(super) fn get_best_sjw(best_tbaud: u32, best_tseg2: u32, sync_jump_width: u16) -> u32 {
    let mut best_sjw = 0;
    let mut best_error = 10000.0;

    for temp_sjw in 1..=best_tseg2 {
        let temp_sync_jump_width = (temp_sjw as f32 * 10000.0) / best_tbaud as f32;
        let error = (temp_sync_jump_width - sync_jump_width as f32).abs();

        if best_error > error {
            best_sjw = temp_sjw;
            best_error = error;
        }
    }
    best_sjw
}

pub(super) struct BitTiming {
    pub(super) brp: u16,
    pub(super) sjw: u8,
    pub(super) tseg1: u8,
    pub(super) tseg2: u8,
}

pub(super) fn calculate_bit_timing(
    module_freq: f32,
    baud_rate: u32,
    sample_point: u16,
    sjw: u16,
) -> BitTiming {
    /* Set values into node */
    let best = get_best_baud_rate(
        NBTP_NBRP_MSK,
        NBTP_NTSEG1_MSK,
        NBTP_NTSEG2_MSK,
        module_freq,
        baud_rate,
    );

    let (best_tseg1, best_tseg2) =
        get_best_sample_point(NBTP_NTSEG1_MSK, NBTP_NTSEG2_MSK, best.tbaud, sample_point);
    let best_sjw = get_best_sjw(best.tbaud as _, best_tseg2 as _, sjw);

    // TODO check this smell, why 1 is subtracted from these values?
    BitTiming {
        brp: best.brp as u16 - 1,
        sjw: best_sjw as u8 - 1,
        tseg1: best_tseg1 as u8 - 1,
        tseg2: best_tseg2 as u8 - 1,
    }
}

pub(super) fn calculate_fast_bit_timing(
    module_freq: f32,
    baud_rate: u32,
    sample_point: u16,
    sjw: u16,
) -> BitTiming {
    /* Set values into node */
    let best = get_best_baud_rate(
        DBTP_DBRP_MSK,
        DBTP_DTSEG1_MSK,
        DBTP_DTSEG2_MSK,
        module_freq,
        baud_rate,
    );

    let (best_tseg1, best_tseg2) =
        get_best_sample_point(DBTP_DTSEG1_MSK, DBTP_DTSEG2_MSK, best.tbaud, sample_point);
    let best_sjw = get_best_sjw(best.tbaud as _, best_tseg2 as _, sjw);

    // TODO check this smell, why 1 is subtracted from these values?
    BitTiming {
        brp: best.brp as u16 - 1,
        sjw: best_sjw as u8 - 1,
        tseg1: best_tseg1 as u8 - 1,
        tseg2: best_tseg2 as u8 - 1,
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_get_best_baudrate() {
        let module_freq = 80000000.0;
        let baudrate = 500000;
        let best = get_best_baud_rate(
            NBTP_NBRP_MSK,
            NBTP_NTSEG1_MSK,
            NBTP_NTSEG2_MSK,
            module_freq,
            baudrate,
        );

        assert_eq!(best.tbaud, 20);
        assert_eq!(best.brp, 8);
    }

    #[test]
    fn test_get_best_sample_point() {
        let sample_point = 8000;
        let best_tbaud = 20;

        let (best_tseg1, best_tseg2) =
            get_best_sample_point(NBTP_NTSEG1_MSK, NBTP_NTSEG2_MSK, best_tbaud, sample_point);

        assert_eq!(best_tseg1, 15);
        assert_eq!(best_tseg2, 4);
    }

    #[test]
    fn test_get_best_sjw() {
        let best_tboud = 20;
        let best_tseg2 = 4;
        let sync_jump_width = 3;
        let best_sjw = get_best_sjw(best_tboud, best_tseg2, sync_jump_width);

        assert_eq!(best_sjw, 1);
    }
}
