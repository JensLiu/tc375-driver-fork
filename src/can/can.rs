// CanNode trait impl to be moved on a separate file (annabo)
pub trait ACanModule {
    fn node_id(&self) -> u8;
    fn is_enabled(&self) -> bool;
    fn is_suspended(&self) -> bool;

    fn enable_module(&self);
    fn disable_module(&self);
    fn reset_module(&self);
    fn init_module(&self);
    //fn set_clock_source(&self, clock_select: ClockSelect, clock_source: ClockSource);
}

pub struct CanModule0 {
    inner: tc37x_pac::can0::Can0,
}

// TODO (annabo) use tc37x_pac::can1::Can1; impl CanModule for Can1
use crate::scu;

#[cfg(feature = "log")]
#[cfg(target_arch = "tricore")]
use defmt::println;
use embedded_can::nb::Can;

impl ACanModule for CanModule0 {
    fn node_id(&self) -> u8 {
        0
    }

    #[inline]
    fn is_enabled(&self) -> bool {
        !unsafe { self.inner.clc().read() }.diss().get()
    }

    fn is_suspended(&self) -> bool {
        unsafe { self.inner.ocs().read() }.sussta().get()
    }
   
    fn enable_module(&self) {
        let passw = scu::wdt::get_cpu_watchdog_password();

        #[cfg(feature = "log")]
        println!("enable module watchdog passw: {:x}", passw);

        scu::wdt::clear_cpu_endinit_inline(passw);

        unsafe { self.inner.clc().modify_atomic(|r| r.disr().set(false)) };
        while !self.is_enabled() {}

        scu::wdt::set_cpu_endinit_inline(passw);
    }

    fn disable_module(&self) {
        let passw = scu::wdt::get_cpu_watchdog_password();

        scu::wdt::clear_cpu_endinit_inline(passw);
        unsafe { self.inner.clc().modify(|r| r.disr().set(true)) };

        while self.is_enabled() {}

        scu::wdt::set_cpu_endinit_inline(passw);
    }

    fn reset_module(&self) {
        let passw = scu::wdt::get_cpu_watchdog_password();

        scu::wdt::clear_cpu_endinit_inline(passw);
        unsafe { self.inner.krst0().modify(|r| r.rst().set(true)) };
        unsafe { self.inner.krst1().modify(|r| r.rst().set(true)) };
        scu::wdt::set_cpu_endinit_inline(passw);

        while !unsafe { self.inner.krst0().read() }.rststat().get() {}

        scu::wdt::clear_cpu_endinit_inline(passw);
        unsafe { self.inner.krstclr().init(|r| r.clr().set(true)) };
        scu::wdt::set_cpu_endinit_inline(passw);
    }

    // fn set_clock_source(&self, clock_select: ClockSelect, clock_source: ClockSource) {
    //     let mut mcr = unsafe { tc37x_pac::CAN0.mcr().read() };
    //     mcr = mcr.ccce().set(true).ci().set(true);

    //     unsafe { tc37x_pac::CAN0.mcr().write(mcr) };

    //     mcr = mcr.clksel(clock_select.into()).set(clock_source.into());
    //     unsafe { tc37x_pac::CAN0.mcr().write(mcr) };

    //     mcr = mcr.ccce().set(false).ci().set(false);
    //     unsafe { tc37x_pac::CAN0.mcr().write(mcr) };
    // }
    fn init_module(&self) {
        if !self.is_enabled() {
            #[cfg(feature = "log")]
            defmt::debug!("module was disabled, enabling it");
            self.enable_module();
        }
    }
}

impl CanModule0{

    pub fn new() -> Self {
        let m = Self {
            inner: tc37x_pac::CAN0,
        }; 
        m.enable_module(); 
        m
        

    }

}