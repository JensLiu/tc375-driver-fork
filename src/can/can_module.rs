use super::can_node::{Node, NodeConfig, NodeId};

use crate::util::wait_nop_cycles;
use crate::{pac, scu};
use core::marker::PhantomData;

use pac::hidden::CastFrom;

#[derive(Clone, Copy)]
pub enum ModuleId {
    Can0,
    Can1,
}

#[derive(Default)]
pub struct ModuleConfig {}

// Type states for Module
pub struct Disabled;
pub struct Enabled;

pub struct Module<Reg, State = Disabled> {
    nodes_taken: [bool; 4],
    _phantom: PhantomData<(Reg, State)>,
}

macro_rules! impl_can_module {
    ($reg:path, $($m:ident)::+, $Reg:ty, $id: expr) => {
        impl Module<$Reg, Disabled> {
            pub const fn new() -> Self {
                Self {
                    nodes_taken: [false; 4],
                    _phantom: PhantomData,
                }
            }

            fn is_enabled(&self) -> bool {
                !unsafe { $reg.clc().read() }.diss().get()
            }

            pub fn enable(self) -> Module<$Reg, Enabled> {
                let passw = scu::wdt::get_cpu_watchdog_password();
                scu::wdt::clear_cpu_endinit_inline();

                unsafe { $reg.clc().modify_atomic(|r| r.disr().set(false)) };
                while !self.is_enabled() {}

                scu::wdt::set_cpu_endinit_inline();

                Module::<$Reg, Enabled> {
                    nodes_taken: [false; 4],
                    _phantom: PhantomData,
                }
            }
        }

        impl Module<$Reg, Enabled> {
            // This method is private to prevent the user from creating a new instance of an enabled module
            const fn new() -> Self {
                Self {
                    nodes_taken: [false; 4],
                    _phantom: PhantomData,
                }
            }

            pub fn take_node(&mut self, node_id: NodeId, cfg: NodeConfig) -> Option<Node<$($m)::+::N, $Reg>> {
                let node_index = match node_id {
                    NodeId::Node0 => 0,
                    NodeId::Node1 => 1,
                    NodeId::Node2 => 2,
                    NodeId::Node3 => 3,
                };

                if self.nodes_taken[node_index] {
                    return None;
                }

                self.nodes_taken[node_index] = true;

                Node::<$($m)::+::N, $Reg>::new(self, node_id, cfg).ok()
            }

            pub fn id(&self) -> ModuleId {
                $id
            }

            pub(crate) fn set_clock_source(
                &self,
                clock_select: ClockSelect,
                clock_source: ClockSource,
            ) -> Result<(), ()> {
                use $($m)::+::mcr::{Ccce, Ci, Clksel0, Clksel1, Clksel2, Clksel3};

                let mcr = self.read_mcr();

                // Enable CCCE and CI
                let mcr = mcr
                    .ccce()
                    .set($($m)::+::mcr::Ccce::CONST_11)
                    .ci()
                    .set($($m)::+::mcr::Ci::CONST_11);
                self.write_mcr(mcr);

                // Select clock
                let clock_source: u8 = clock_source.into();

                let mcr = match clock_select.0 {
                    0 => mcr.clksel0().set(Clksel0::cast_from(clock_source.into())),
                    1 => mcr.clksel1().set(Clksel1::cast_from(clock_source.into())),
                    2 => mcr.clksel2().set(Clksel2::cast_from(clock_source.into())),
                    3 => mcr.clksel3().set(Clksel3::cast_from(clock_source.into())),
                    _ => unreachable!(),
                };

                self.write_mcr(mcr);

                // Disable CCCE and CI
                let mcr = mcr.ccce().set(Ccce::CONST_00).ci().set(Ci::CONST_00);
                self.write_mcr(mcr);

                // TODO Is this enough or we need to wait until actual_clock_source == clock_source
                // Wait for clock switch
                 wait_nop_cycles(10);

                // Check if clock switch was successful
                let mcr = self.read_mcr();

                let actual_clock_source = match clock_select.0 {
                    0 => mcr.clksel0().get().0,
                    1 => mcr.clksel1().get().0,
                    2 => mcr.clksel2().get().0,
                    3 => mcr.clksel3().get().0,
                    _ => unreachable!(),
                };

                if actual_clock_source != clock_source {
                    return Err(());
                }

                Ok(())
            }

            pub(crate) fn registers(&self) -> &$Reg {
                &$reg
            }

            fn read_mcr(&self) -> $($m)::+::Mcr {
                unsafe { $reg.mcr().read() }
            }

            fn write_mcr(&self, mcr: $($m)::+::Mcr) {
                unsafe { $reg.mcr().write(mcr) }
            }

            pub(crate) fn ram_base_address(&self) -> usize {
                $reg.0 as usize
            }
        }
    };
}

impl_can_module!(pac::CAN0, pac::can0, pac::can0::Can0, ModuleId::Can0);
impl_can_module!(pac::CAN1, pac::can1, pac::can1::Can1, ModuleId::Can1);

pub(crate) struct ClockSelect(u8);

impl From<NodeId> for ClockSelect {
    fn from(node_id: NodeId) -> Self {
        Self(node_id.into())
    }
}

#[derive(Default, Clone, Copy)]
pub enum ClockSource {
    Asynchronous,
    Synchronous,
    #[default]
    Both,
}

impl From<ClockSource> for u8 {
    fn from(x: ClockSource) -> Self {
        match x {
            ClockSource::Asynchronous => 1,
            ClockSource::Synchronous => 2,
            ClockSource::Both => 3,
        }
    }
}
