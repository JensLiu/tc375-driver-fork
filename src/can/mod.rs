// TODO Remove this once the module is more complete
#![allow(dead_code)]

mod baud_rate;
mod can_module;
mod can_node;
mod frame;
mod internals;
pub mod msg;
mod reg;

pub use baud_rate::{AutoBitTiming, BitTimingConfig, DataBitTiming, NominalBitTiming};
pub use can_module::{Module, ModuleConfig, ModuleId};
pub use can_node::{DataFieldSize, Node, NodeConfig, NodeId, TxConfig, TxMode};
pub use frame::Frame;
pub use msg::MessageId;
