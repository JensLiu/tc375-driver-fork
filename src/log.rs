//! Logging macros and utilities.

// Log with defmt
#[cfg(all(not(feature = "log_with_env_logger"), feature = "log_with_defmt"))]
pub use tricore_log::*;

// No log system
#[cfg(all(not(feature = "log_with_defmt"), not(feature = "log_with_env_logger")))]
pub use crate::{
    no_log_debug as debug, no_log_error as error, no_log_info as info, no_log_trace as trace,
    no_log_warn as warn,
};

// Log with log crate
#[cfg(all(not(feature = "log_with_defmt"), feature = "log_with_env_logger"))]
pub use ::log::{debug, error, info, trace, warn};

// link with defmt implementation
#[cfg(feature = "log_with_defmt")]
extern crate defmt_rtt;

#[cfg(feature = "log_with_defmt")]
mod tricore_log {
    pub use defmt::debug;
    pub use defmt::error;
    pub use defmt::info;
    pub use defmt::trace;
    pub use defmt::warn;
}

#[macro_export]
macro_rules! no_log_debug {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_trace {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_info {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_warn {
    ($($ignored:tt)*) => {};
}

#[macro_export]
macro_rules! no_log_error {
    ($($ignored:tt)*) => {};
}

/// A wrapper around a slice of bytes that implements `Display` to print the
/// bytes as a hex string.
///
/// ```
/// use bw_r_drivers_tc37x::log::HexSlice;
/// let bytes = &[1u8, 2, 3, 4];
/// assert_eq!(format!("{}", HexSlice::from(bytes)), "01020304");
/// ```
pub struct HexSlice<'a>(&'a [u8]);

impl core::fmt::Display for HexSlice<'_> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        for byte in self.0 {
            write!(f, "{:02x}", byte)?;
        }
        Ok(())
    }
}

#[cfg(feature = "log_with_defmt")]
impl defmt::Format for HexSlice<'_> {
    fn format(&self, f: defmt::Formatter) {
        for byte in self.0 {
            defmt::write!(f, "{:02x}", byte);
        }
    }
}

impl<'a, const N: usize> From<&'a [u8; N]> for HexSlice<'a> {
    fn from(value: &'a [u8; N]) -> Self {
        HexSlice(value.as_slice())
    }
}
impl<'a> From<&'a [u8]> for HexSlice<'a> {
    fn from(value: &'a [u8]) -> Self {
        HexSlice(value)
    }
}
