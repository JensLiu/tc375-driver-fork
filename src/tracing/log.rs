#![allow(clippy::expect_used)]

use super::TraceGuard;
use crate::tracing::{LoadModifyStoreEntry, ReadEntry, ReportEntry, WriteEntry};
use std::any::Any;
use std::collections::{HashMap, VecDeque};
use std::fmt::{Debug, Display, Formatter, Write};
use std::sync::{Arc, Mutex, MutexGuard};

struct ReadFifoEntry {
    addr: usize,
    len: usize,
    val: u64,
}

#[derive(Default)]
struct ReadFifo(VecDeque<ReadFifoEntry>);

#[derive(Default)]
struct SharedData {
    log: Log,
    read_fifo: ReadFifo,
}

struct Reporter {
    shared_data: Arc<Mutex<SharedData>>,
}

pub struct Report {
    shared_data: Arc<Mutex<SharedData>>,
    _guard: TraceGuard,
}

impl Default for Report {
    fn default() -> Self {
        Self::new()
    }
}

impl Report {
    pub fn new() -> Self {
        let data = Arc::new(Mutex::new(SharedData::default()));
        let reporter = Reporter {
            shared_data: Arc::clone(&data),
        };
        let guard = TraceGuard::new(reporter);
        Self {
            shared_data: data,
            _guard: guard,
        }
    }

    pub fn take_log(&self) -> Log {
        let mut g = self.shared_data();
        let len = g.log.0.len();
        Log(g.log.0.drain(0..len).collect())
    }

    pub fn expect_read(&self, addr: usize, len: usize, val: u64) {
        self.shared_data()
            .read_fifo
            .0
            .push_front(ReadFifoEntry { addr, len, val })
    }

    pub fn comment(&self, s: impl Into<String>) {
        self.shared_data().log.0.push(LogEntry::Comment(s.into()));
    }

    fn shared_data(&self) -> MutexGuard<SharedData> {
        self.shared_data.lock().unwrap()
    }
}

impl Reporter {
    fn push(&self, report: ReportEntry) {
        self.shared_data().log.0.push(LogEntry::ReportEntry(report));
    }

    fn shared_data(&self) -> MutexGuard<SharedData> {
        self.shared_data.lock().unwrap()
    }
}

impl super::Reporter for Reporter {
    fn read_volatile(&self, addr: usize, len: usize) -> u64 {
        let entry = self
            .shared_data()
            .read_fifo
            .0
            .pop_front()
            .expect("Unexpected read");

        if entry.addr == addr && entry.len == len {
            let val = entry.val;
            self.push(ReportEntry::Read(ReadEntry { addr, len, val }));
            val
        } else {
            panic!("Unexpected read at address 0x{:08X} and len {}", addr, len)
        }
    }

    fn write_volatile(&self, addr: usize, len: usize, val: u64) {
        self.push(ReportEntry::Write(WriteEntry { addr, len, val }));
    }

    fn load_modify_store(&self, addr: usize, val: u64) {
        self.push(ReportEntry::LoadModifyStore(LoadModifyStoreEntry {
            addr,
            val,
        }));
    }
}

impl Drop for Reporter {
    fn drop(&mut self) {
        if !self.shared_data().read_fifo.0.is_empty() {
            panic!("More read where expected");
        }
    }
}

#[derive(Default, Debug, PartialEq)]
pub struct Log(Vec<LogEntry>);

#[derive(Debug, PartialEq)]
enum LogEntry {
    Comment(String),
    ReportEntry(ReportEntry),
}

impl Display for Log {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for entry in &self.0 {
            match entry {
                LogEntry::ReportEntry(ReportEntry::Read(x)) => {
                    write!(f, "r    0x{:08X} {:02} 0x{:08X}", x.addr, x.len, x.val);
                }
                LogEntry::ReportEntry(ReportEntry::Write(x)) => {
                    write!(f, "w    0x{:08X} {:02} 0x{:08X}", x.addr, x.len, x.val);
                }
                LogEntry::ReportEntry(ReportEntry::LoadModifyStore(x)) => {
                    let mask = (x.val >> 32);
                    let val = (x.val & 0xFFFFFFFF);
                    write!(f, "ldms 0x{:08X} 0x{:08X} 0x{:08X}", x.addr, mask, val);
                }
                LogEntry::Comment(s) => {
                    write!(f, "# {s}");
                }
            }

            f.write_char('\n');
        }

        Ok(())
    }
}
