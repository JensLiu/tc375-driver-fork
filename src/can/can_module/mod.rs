mod service_request;

use super::can_node::{Node, NodeConfig};

use crate::can::Priority;
use crate::can::Tos;
use crate::can::{InterruptLine, NodeId};
use crate::util::wait_nop_cycles;
use crate::{pac, scu};
use core::marker::PhantomData;
use pac::hidden::CastFrom;

pub trait ModuleId {}

pub struct Module0;
impl ModuleId for Module0 {}

pub struct Module1;
impl ModuleId for Module1 {}

// Type states for Module
pub struct Disabled;
pub struct Enabled;

pub struct Module<ModuleId, Reg, State> {
    nodes_taken: [bool; 4],
    _phantom: PhantomData<(ModuleId, Reg, State)>,
}

impl<ModuleId, Reg> Module<ModuleId, Reg, Disabled> {
    /// Create a new (disabled) CAN module
    pub fn new(_module_id: ModuleId) -> Self {
        Self {
            nodes_taken: [false; 4],
            _phantom: PhantomData,
        }
    }
}

macro_rules! impl_can_module {
    ($module_reg:path, $($m:ident)::+, $ModuleReg:ty, $ModuleId: ty) => {
        impl Module<$ModuleId, $ModuleReg, Disabled> {
            fn is_enabled(&self) -> bool {
                !unsafe { $module_reg.clc().read() }.diss().get()
            }

            /// Enable the CAN module
            pub fn enable(self) -> Module<$ModuleId, $ModuleReg, Enabled> {
                scu::wdt::clear_cpu_endinit_inline();

                unsafe { $module_reg.clc().modify_atomic(|r| r.disr().set(false)) };
                while !self.is_enabled() {}

                scu::wdt::set_cpu_endinit_inline();

                Module::<$ModuleId, $ModuleReg, Enabled> {
                    nodes_taken: [false; 4],
                    _phantom: PhantomData,
                }
            }
        }

        impl Module<$ModuleId, $ModuleReg, Enabled> {
            /// Take ownership of a CAN node and configure it
            pub fn take_node<I>(&mut self, node_id: I, cfg: NodeConfig<$ModuleReg, I>) -> Option<Node<$($m)::+::N, $ModuleReg>> where I: NodeId {
                let node_index = node_id.as_index();

                // Check if node is already taken, return None if it is
                if self.nodes_taken[node_index] {
                    return None;
                }

                // Mark node as taken
                self.nodes_taken[node_index] = true;

                // Create node
                Node::<$($m)::+::N, $ModuleReg>::new(self, node_id, cfg).ok()
            }

            pub(crate) fn set_clock_source(
                &self,
                clock_select: ClockSelect,
                clock_source: ClockSource,
            ) -> Result<(), ()> {
                use $($m)::+::mcr::{Ccce, Ci, Clksel0, Clksel1, Clksel2, Clksel3};

                let mcr = self.read_mcr();

                // Enable CCCE and CI
                let mcr = mcr
                    .ccce()
                    .set($($m)::+::mcr::Ccce::CONST_11)
                    .ci()
                    .set($($m)::+::mcr::Ci::CONST_11);
                self.write_mcr(mcr);

                // Select clock
                let clock_source: u8 = clock_source.into();

                let mcr = match clock_select.0 {
                    0 => mcr.clksel0().set(Clksel0::cast_from(clock_source.into())),
                    1 => mcr.clksel1().set(Clksel1::cast_from(clock_source.into())),
                    2 => mcr.clksel2().set(Clksel2::cast_from(clock_source.into())),
                    3 => mcr.clksel3().set(Clksel3::cast_from(clock_source.into())),
                    _ => unreachable!(),
                };

                self.write_mcr(mcr);

                // Disable CCCE and CI
                let mcr = mcr.ccce().set(Ccce::CONST_00).ci().set(Ci::CONST_00);
                self.write_mcr(mcr);

                // TODO Is this enough or we need to wait until actual_clock_source == clock_source
                // Wait for clock switch
                 wait_nop_cycles(10);

                // Check if clock switch was successful
                let mcr = self.read_mcr();

                let actual_clock_source = match clock_select.0 {
                    0 => mcr.clksel0().get().0,
                    1 => mcr.clksel1().get().0,
                    2 => mcr.clksel2().get().0,
                    3 => mcr.clksel3().get().0,
                    _ => unreachable!(),
                };

                if actual_clock_source != clock_source {
                    return Err(());
                }

                Ok(())
            }

            pub(crate) fn set_interrupt(
                &self,
                line: InterruptLine,
                priority: Priority,
                tos: Tos,
            ) {
                <$ModuleId>::service_request(line).enable(priority, tos)
            }

            pub(crate) fn registers(&self) -> &$ModuleReg {
                &$module_reg
            }

            fn read_mcr(&self) -> $($m)::+::Mcr {
                unsafe { $module_reg.mcr().read() }
            }

            fn write_mcr(&self, mcr: $($m)::+::Mcr) {
                unsafe { $module_reg.mcr().write(mcr) }
            }

            pub(crate) fn ram_base_address(&self) -> usize {
                $module_reg.0 as usize
            }
        }
    };
}

impl_can_module!(pac::CAN0, pac::can0, pac::can0::Can0, Module0);
impl_can_module!(pac::CAN1, pac::can1, pac::can1::Can1, Module1);

pub(crate) struct ClockSelect(pub(crate) u8);

impl<T> From<T> for ClockSelect
where
    T: NodeId,
{
    fn from(value: T) -> Self {
        ClockSelect(value.as_index() as u8)
    }
}

#[derive(Default, Clone, Copy)]
pub enum ClockSource {
    Asynchronous,
    Synchronous,
    #[default]
    Both,
}

impl From<ClockSource> for u8 {
    fn from(x: ClockSource) -> Self {
        match x {
            ClockSource::Asynchronous => 1,
            ClockSource::Synchronous => 2,
            ClockSource::Both => 3,
        }
    }
}
