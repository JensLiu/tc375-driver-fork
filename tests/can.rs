use tc37x_driver::tracing::log::Report;
use tc37x_driver::can::Module;
use tc37x_driver::can::Module0;

#[test]
fn test_can_module_enable(){
    let report = Report::new();
    let can_module = Module::new(Module0);

    // wtdcpu0con0 read
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_1_1);
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_1_1);
    // wtdcpu0con0 write
    // wtdcpu0con0 read
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_0_1);

    // wtdcpu0con0 write

    // clc read
    report.expect_read(0xF0208000, 4, 0b00000000000000000_00000000_0000_0_0_0);

    // wtdcpu0con0 read
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_1_0);
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_1_0);

    // wtdcpu0con0 write
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_0_1);
    report.expect_read(0xF003624C, 4, 0b1111111111111100_00000000111100_0_1);



    let mut can_module = can_module.enable();

    insta::assert_snapshot!(report.take_log());
}
