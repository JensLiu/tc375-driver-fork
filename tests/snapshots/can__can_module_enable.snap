---
source: tests/can.rs
assertion_line: 36
expression: report.take_log()
---
r    0xF003624C 04 0xFFFC00F3
r    0xF003624C 04 0xFFFC00F3
r    0xF003624C 04 0xFFFC00F3
w    0xF003624C 04 0xFFFC000D
r    0xF003624C 04 0xFFFC00F1
w    0xF003624C 04 0xFFFC000E
ldms 0xF0208000 0x00000001 0x00000002
r    0xF0208000 04 0x00000000
r    0xF003624C 04 0xFFFC00F2
r    0xF003624C 04 0xFFFC00F2
r    0xF003624C 04 0xFFFC00F2
w    0xF003624C 04 0xFFFC000D
r    0xF003624C 04 0xFFFC00F1
w    0xF003624C 04 0xFFFC000F
r    0xF003624C 04 0xFFFC00F1
